### PR TITLE
[5.2] Use league/uri for generating URIs

### PR DESCRIPTION
**Pull request by rkgrep**
Gives much more control on constructing URIs

### CODE REVIEW

**Comment by rkgrep** [comment]
See #10923 for unit test fix

**Comment by lucasmichot** [comment]
The drawback to that is that a lot of PHP extensions dependencies are added

**Comment by lucasmichot** [comment]
Unrelated / @GrahamCampbell any reason why the StyleCI status of this PR appears (https://styleci.io/analyses/8n5evz) and then is removed ?

**Comment by rkgrep** [comment]
@lucasmichot `ext-fileinfo` is [required](https://github.com/thephpleague/flysystem/blob/master/composer.json#L20) anyway by `league/flysystem`, but not mentioned in laravel.

I fixed CS right after StyleCI reported error and made a force update. Maybe that is the reason.

**Comment by lucasmichot** [comment]
Well `league/flysystem` is suggested, not required...

**Comment by rkgrep** [comment]
It is required, actually https://github.com/laravel/framework/blob/5.1/composer.json#L25

**Comment by lucasmichot** [comment]
My bad, @rkgrep you're right

**Comment by GrahamCampbell** [comment]
> any reason why the StyleCI status of this PR appears 

GitHub's API was down momentarily so StyleCI failed to report the status to GitHub. I've closed and re-opened to re-trigger StyleCI.

File: composer.json [hunk 1]
```diff
@@ -18,11 +18,14 @@
         "php": ">=5.5.9",
         "ext-mbstring": "*",
         "ext-openssl": "*",
+        "ext-intl" : "*",
```
Comment (GrahamCampbell) [inline composer.json]: don't add this

File: composer.json [hunk 2]
```diff
@@ -18,11 +18,14 @@
...
+        "ext-fileinfo": "*",
```
Comment (GrahamCampbell) [inline composer.json]: don't add this

**Comment by rkgrep** [comment]
@lucasmichot Never mind :smiley: 
I face this dependency missing in framework every time while deploying on a new server and always forget to make a PR.

File: composer.json (hunk 1 above)
Comment (GrahamCampbell) [inline composer.json]: league/uri does this for us - don't add them again

File: src/Illuminate/Routing/UrlGenerator.php [hunk 3]
```diff
@@ -650,6 +653,17 @@ protected function trimUrl($root, $path, $tail = '')
     }
 
     /**
+     * Create Uri object from string.
+     *
+     * @param  $uri
```
Comment (GrahamCampbell) [inline src/Illuminate/Routing/UrlGenerator.php]: cs

File: src/Illuminate/Routing/composer.json [hunk 4]
```diff
@@ -15,12 +15,15 @@
     ],
     "require": {
         "php": ">=5.5.9",
+        "ext-intl" : "*",
```
Comment (GrahamCampbell) [inline src/Illuminate/Routing/composer.json]: no

File: src/Illuminate/Routing/composer.json [hunk 5]
```diff
@@ -15,12 +15,15 @@
...
+        "ext-fileinfo": "*",
```
Comment (GrahamCampbell) [inline src/Illuminate/Routing/composer.json]: no

**Comment by GrahamCampbell** [comment]
Maybe we should hold off until L5.3?

**Comment by rkgrep** [comment]
> Maybe we should hold off until L5.3?

I would like to add to change `@return` value to `string|\Psr\Http\Message\UriInterface` in contract then

**Comment by taylorotwell** [comment]
This isn't actually helping us much.

**Comment by rkgrep** [comment]
What about an internal package for url handling as I proposed in #10920?

**Comment by taylorotwell** [comment]
It's not bad. Main concern would be breaking changes.

**Comment by rkgrep** [comment]
It is not a great break, as the class must have `__toString()` method according to PSR standard.

**Comment by rkgrep** [comment]
@taylorotwell Haven't you changed your mind about this for 5.3? It would be great to have customisable URIs out of the box.

**Comment by peter-mw** [comment]
one thing, seems league/uri requires the `ext-intl` extension which is not commonly installed

i get this error on shared hosting environment
 
` - league/uri 4.2.x-dev requires ext-intl * -> the requested PHP extension intl is missing from your system.
`

**Comment by driesvints** [comment]
@peter-mw which PHP version are you running? Should come bundled as of PHP 5.3. Could be that it just needs to be turned on.

**Comment by peter-mw** [comment]
@driesvints i am running this on PHP 5.6.14 on cPanel shared hosting, maybe is hosting specific issue

**Comment by GrahamCampbell** [comment]
Maybe your shared host disabled that extension?

**Comment by rkgrep** [comment]
I think that missing extension is not a problem in this case, because it is usually bundled in PHP installations as @driesvints said.
However if it is not present, it may be requested to hosting support to enable it.

**Pull request closed without merging**